### ISSUE TITLE
Yarn PnP compatibility

### ISSUE THREAD

**Issue by notpeelz** [esbuild-kit/tsx#166]
### Bug description

<details>
  <summary>Running tsx in a PnP-enabled project results in this error:</summary>

```
$ yarn tsx
node:fs:1596
  handleErrorFromBinding(ctx);
  ^

Error: ENOTDIR: not a directory, stat '/home/peelz/testproject/.yarn/cache/tsx-npm-3.12.1-6df77f29e6-7251ad550b.zip/node_modules/tsx/dist/loader.js'
    at statSync (node:fs:1596:3)
    at tryStatSync (node:internal/modules/esm/resolve:176:13)
    at finalizeResolution (node:internal/modules/esm/resolve:316:17)
    at moduleResolve (node:internal/modules/esm/resolve:916:10)
    at defaultResolve (node:internal/modules/esm/resolve:1124:11)
    at nextResolve (node:internal/modules/esm/loader:163:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:841:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:424:18)
    at ESMLoader.import (node:internal/modules/esm/loader:525:22)
    at initializeLoader (node:internal/process/esm_loader:75:58) {
  errno: -20,
  syscall: 'stat',
  code: 'ENOTDIR',
  path: '/home/peelz/testproject/.yarn/cache/tsx-npm-3.12.1-6df77f29e6-7251ad550b.zip/node_modules/tsx/dist/loader.js'
}

Node.js v18.12.1
```
</details>

However, this seems to be fixed by https://github.com/nodejs/node/pull/43772 (yay!).
If you want to try a working test environment of tsx + Yarn PnP, you can use this demo project: https://github.com/notpeelz/tsx-yarn-pnp-demo.

### Reproduction

On the current stable build of Node 18.x/19.x, you can reproduce the issue like this:
1. Install Yarn: `npm i -g yarn`
2. Create a new project: `mkdir testproject && cd testproject && yarn init -p -2`
3. Install tsx: `yarn add -D tsx`
4. Run tsx: `yarn tsx`

### Environment

```shell
System:
  OS: Linux 5.15 Debian GNU/Linux bookworm/sid
  CPU: (24) x64 AMD Ryzen 9 3900X 12-Core Processor
  Memory: 11.24 GB / 15.58 GB
  Container: Yes
  Shell: 5.9 - /usr/bin/zsh
Binaries:
  Node: 18.12.1 - ~/.nvm/versions/node/v18.12.1/bin/node
  Yarn: 3.3.1 - ~/.nvm/versions/node/v18.12.1/bin/yarn
  npm: 8.19.2 - ~/.nvm/versions/node/v18.12.1/bin/npm
```

### Can you contribute a fix?

- [ ] I’m interested in opening a pull request for this issue.

**Comment by privatenumber** [comment]
I'm skeptical about this because neither yarn or tsx doesn't appear in your error stack. If there's a yarn pnp loader that resolves the tsx loader, it should appear in the stack.

And, as I understand, the reproduction you provided for "the current stable build of Node 18.x/19.x" isn't sufficient because the PR you linked isn't released yet. In a reproduction, it should be possible to attain a pass case without external changes.

I'm not sure if I will debug an unreleased version of Node.js, but I tried following your reproduction steps but a `.pnp.loader.mjs` was not generated for me.

**Comment by notpeelz** [comment]
There's nothing to debug or fix on your part. I spent hours tracking down this issue and debugging the problem, and I believe this issue will be fixed once https://github.com/nodejs/node/pull/43772 gets backported to the 18.x/19.x branches.

I tested the patch myself by manually merging it into the current 18.x branch. If you're curious, you can test it yourself with [the demo project I linked](https://github.com/notpeelz/tsx-yarn-pnp-demo).

I only opened this issue for tracking purposes -- so other people don't have to go down this rabbit hole. I apologize for not making that clear in my original report 😅 

--------

> I'm skeptical about this because neither yarn or tsx appear in your error stack. If there's a yarn pnp loader that resolves the tsx loader, it should appear in the stack.

I believe this stack trace belongs to the process you spawn through the `cli.js` entrypoint: https://github.com/esbuild-kit/tsx/blob/be0478990305fc0052d9c6c982a20d3990658cc2/src/run.ts#L31-L46

It doesn't have any mention of yarn/pnp because the PnP API hasn't been loaded, so it can't resolve files through ZipFS.

When you launch a "bin" package through yarn (e.g `yarn tsx`), Yarn will spawn a Node process with the `--require ./.pnp.cjs`. These parameters don't get passed to child processes, so they use the `NODE_OPTIONS` environment variable instead: https://github.com/yarnpkg/berry/blob/3aac4e8585d2e89453735a6e12db1ff7cbb57c8a/packages/plugin-exec/sources/ExecFetcher.ts#L140-L153

When you spawn your process with `--require .../preflight.cjs --loader .../loader.js`, it also inherits the `NODE_OPTIONS` env var, which results in something like this: `node --require ./.pnp.cjs --require .../preflight.cjs ---loader .../loader.js`.

I don't know the specifics, but it seems that https://github.com/nodejs/node/pull/43772 fixes this issue.
Using this patch, **tsx and Yarn PnP work perfectly together** (as far as I can tell).

**Comment by privatenumber** [comment]
How did you generate `.pnp.loader.mjs `?

**Comment by notpeelz** [comment]
You need to add this to your `.yarnrc.yml`:
```yml
pnpEnableEsmLoader: true
```

**Comment by arcanis** [comment]
It's also generated by default if you or one of the packages in the dependency tree has `type: "module"`

**Comment by slikts** [comment]
I ran into the same issue, and upgrading node to 19.5 solves it.

**Comment by arnamak** [comment]
I'm still struggling with this despite the really helpful information here (thanks everyone!)

It seems like both `tsx` and `yarn` struggle with the eventual TSConfig setup of:

```
module: "NodeNext"
moduleResolution: "bundler" // (available in TS 5.0) "nodenext" doesn't work here either for ^4.9
```

On the `tsx` side, should we expect that we'll be able to declare relative imports without a file extension (which TS requires when ES mode is enabled). E.g.:

```
import { someData } from './data/api';
```

Because, as it stands right now I need to specify:
```
import { someData } from './data/api.js';
```


With 5.0, TS is happy with the extension-less variants, but `tsx` is not. by adding the config options above, but `tsx` fails:

<details><summary>with NODE_OPTIONS="-r .pnp.cjs --loader .pnp.loader.mjs" yarn node --loader tsx src/index.(m)ts</summary>

<p>

```
$ NODE_OPTIONS="-r ./.pnp.cjs --loader ./.pnp.loader.mjs" $YARN node --loader tsx src/happy-ts-unhappy-tsx.mts
C:\Users\arnamak\Projects\tsx-yarn-pnp-demo\.pnp.cjs:10054
  return Object.defineProperties(new Error(message), {
                                 ^

Error: Qualified path resolution failed: we looked for the following paths, but none could be accessed.

Source path: C:\Users\arnamak\Projects\tsx-yarn-pnp-demo\src\data\api
Not found: C:\Users\arnamak\Projects\tsx-yarn-pnp-demo\src\data\api

    at makeError (C:\Users\arnamak\Projects\tsx-yarn-pnp-demo\.pnp.cjs:10054:34)
    at resolveUnqualified (C:\Users\arnamak\Projects\tsx-yarn-pnp-demo\.pnp.cjs:11579:13)
    at resolveRequest (C:\Users\arnamak\Projects\tsx-yarn-pnp-demo\.pnp.cjs:11620:14)
    at Object.resolveRequest (C:\Users\arnamak\Projects\tsx-yarn-pnp-demo\.pnp.cjs:11676:26)
    at resolve$1 (file:///C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/.pnp.loader.mjs:1991:25)
    at nextResolve (node:internal/modules/esm/loader:163:28)
    at u (file:///C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/.yarn/cache/@esbuild-kit-esm-loader-npm-2.5.4-e3eb3e01f1-8f4b4b6470.zip/node_modules/@esbuild-kit/esm-loader/dist/index.js:1:2387)
    at nextResolve (node:internal/modules/esm/loader:163:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:835:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:416:18)

Node.js v19.7.0
```

</p>
</details>


Even still, outside of the purview of `tsx` @arcanis because I see you're in here already, if I were to add those file extensions and live with it, TS does not seem to be happy with any kind of import via a dependency say, `lodash` or `lodash-es`:

<details><summary>Example --traceResolution output</summary>
<p>

```
$ NODE_OPTIONS="-r ./.pnp.cjs --loader ./.pnp.loader.mjs" $YARN tsc -p tsconfig.json --traceResolution
======== Resolving module './es-data/api.mjs' from 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/happy-tsx.mts'. ========
Explicitly specified module resolution kind: 'Bundler'.
Loading module as file / folder, candidate module location 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/es-data/api.mjs', target file types: TypeScript, JavaScript, Declaration, JSON.
File name 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/es-data/api.mjs' has a '.mjs' extension - stripping it.
File 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/es-data/api.mts' exist - use it as a name resolution result.
======== Module name './es-data/api.mjs' was successfully resolved to 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/es-data/api.mts'. ========
======== Resolving module './es-data/api.mjs' from 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/unhappy-ts.mts'. ========
Resolution for module './es-data/api.mjs' was found in cache from location 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src'.
======== Module name './es-data/api.mjs' was successfully resolved to 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/es-data/api.mts'. ========
======== Resolving module 'lodash-es' from 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/unhappy-ts.mts'. ========
Explicitly specified module resolution kind: 'Bundler'.
File 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/package.json' does not exist.
Found 'package.json' at 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/package.json'.
Loading module 'lodash-es' from 'node_modules' folder, target file types: TypeScript, JavaScript, Declaration, JSON.
Directory 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/Users/arnamak/Projects/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/Users/arnamak/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/Users/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/Users/arnamak/Projects/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/Users/arnamak/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/Users/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/node_modules' does not exist, skipping all lookups in it.
======== Module name 'lodash-es' was not resolved. ========
======== Resolving module './data/api' from 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/unhappy-ts-and-tsx.mts'. ========
Explicitly specified module resolution kind: 'Bundler'.
Loading module as file / folder, candidate module location 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/data/api', target file types: TypeScript, JavaScript, Declaration, JSON.
File 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/data/api.ts' exist - use it as a name resolution result.
======== Module name './data/api' was successfully resolved to 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/data/api.ts'. ========
======== Resolving module 'lodash' from 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/unhappy-ts-and-tsx.mts'. ========
Explicitly specified module resolution kind: 'Bundler'.
File 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/package.json' does not exist according to earlier cached lookups.
File 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/package.json' exists according to earlier cached lookups.
Loading module 'lodash' from 'node_modules' folder, target file types: TypeScript, JavaScript, Declaration, JSON.
Directory 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/Users/arnamak/Projects/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/Users/arnamak/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/Users/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/Users/arnamak/Projects/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/Users/arnamak/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/Users/node_modules' does not exist, skipping all lookups in it.
Directory 'C:/node_modules' does not exist, skipping all lookups in it.
======== Module name 'lodash' was not resolved. ========
======== Resolving module './data/api' from 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/happy-ts-unhappy-tsx.mts'. ========
Resolution for module './data/api' was found in cache from location 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src'.
======== Module name './data/api' was successfully resolved to 'C:/Users/arnamak/Projects/tsx-yarn-pnp-demo/src/data/api.ts'. ========
src/unhappy-ts-and-tsx.mts:2:23 - error TS2307: Cannot find module 'lodash' or its corresponding type declarations.

2 import { merge } from 'lodash';
                        ~~~~~~~~

src/unhappy-ts.mts:2:23 - error TS2307: Cannot find module 'lodash-es' or its corresponding type declarations.

2 import { merge } from 'lodash-es';
                        ~~~~~~~~~~~


Found 2 errors in 2 files.

Errors  Files
     1  src/unhappy-ts-and-tsx.mts:2
     1  src/unhappy-ts.mts:2
```

</p>
</details>

<hr />

If I downgrade to TS 4.9, I get similarly frustrating conflicts.

With:

```
"module": "NodeNext",
"moduleResolution": "nodenext"
```

All imports need a `.js` file extension. The difference with this approach is that `tsc` seems happy with the dependency resolution, but `tsx` does not appear to match:

<details><summary>yarn tsx</summary>
<p>

```
$ NODE_OPTIONS="-r ./.pnp.cjs --loader ./.pnp.loader.mjs" $YARN node --loader tsx src/happy-ts-unhappy-tsx.mts
c:\Users\arnamak\Projects\tsx-yarn-pnp-demo\src\happy-ts-unhappy-tsx.mts:2
import { merge } from 'lodash';
         ^

SyntaxError: The requested module 'lodash' does not provide an export named 'merge'
    at ModuleJob._instantiate (node:internal/modules/esm/module_job:123:21)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async ModuleJob.run (node:internal/modules/esm/module_job:189:5)

Node.js v19.7.0
```
</p>
</details>

<hr />

I made a couple of branches for perusal forked off of @notpeelz example:

[TypeScript 4.9](https://github.com/arnamak/tsx-yarn-pnp-demo/tree/TS-4.9)

[TypeScript 5.0](https://github.com/arnamak/tsx-yarn-pnp-demo/tree/typescript-5-beta)

I realize that 5.0 is still in beta so I don't really _expect_ support for it right now, I'm just posting this here because I've had a _week_ with this stuff and this seemed like the least noisy/relevant place.

<hr />

# tl;dr

- Should `tsx` be able to support relative imports without `js` file extensions?
- Will `yarn` patch 5.0 to support the `"bundler"` moduleResolution flag?

Thank you all for your contributions to the current cluster@#%$ of JavaScript 😄 I really can't express enough respect for the lot of you, despite these struggles all of the tooling you work on is fantastic and it is an inspiring brand of OSS.

**Comment by privatenumber** [comment]
You might be conflating tsx issues with yarns.

tsx aims to handle any TypeScript code the way the TypeScript Compiler. So test tsx using npm or pnpm (instead of yarn) and if something works with TypeScript Compiler but not with tsx, you can file a bug report.

tsx should add prepare to add TypeScript 5 features but not sure when. Definitely shouldn't release TS 5 beta features.

---

Personal advice: I have had a much better experience using [pnpm](https://pnpm.io/) over yarn and highly recommend others to migrate.

**Comment by arcanis** [comment]
> Even still, outside of the purview of `tsx` @arcanis because I see you're in here already, if I were to add those file extensions and live with it, TS does not seem to be happy with any kind of import via a dependency say, `lodash` or `lodash-es`:

Your post has a little _too much_ information, it's hard for me to give good insight with just a glance 🥲

Generally:

- The original issue is fixed, now that Node.js has landed the bugfix (make sure Node is up-to-date).
- I've been using the `bundler` resolution with `tsc` just fine in some of my own projects.
- To confirm it's a problem with PnP, just run `yarn config set nodeLinker node-modules && yarn`. If you still have errors, the problem comes from `tsx` and/or Node.js. If you don't, it might come from PnP, in which case you should open an issue on the Yarn repository.

In the meantime, since the original issue is solved, I think this issue should be closed.

**Comment by khokm** [comment]
Why is it closed?! The problem isn't solved.
@arcanis Yes, it works with `nodeLinker: node_modules`. But this is still a tsx problem, not Yarn.
I created a two files: `index.ts` and `index.js`. They both just import some package.

`yarn esbuild --platform=node --bundle src/index.ts` works fine.
`yarn node src/index.js` works too.
And `yarn tsx src/index.ts` return the error mentioned in OP post.

**Comment by privatenumber** [comment]
It's actually a Node.js limitation as you can tell from your fix.

I just documented this limitation in https://github.com/esbuild-kit/tsx/commit/2a83619ab2b14df0f52a338288520c366a5ebc2f so it's easier for others to figure out in the future.
